### PR TITLE
Adds lazy resource property

### PR DESCRIPTION
**Pull request by shubham1g5**
Multimedia Changes - 

1. Adds a lazy `Resource` property
2. Better classify `MissingMediaException` using `MissingMediaExceptionType`

cross-request: https://github.com/dimagi/commcare-android/pull/2243

### CODE REVIEW

File: src/main/java/org/commcare/resources/model/ResourceTable.java [hunk 1]
```diff
@@ -1175,4 +1190,9 @@ public void setMissingResources(SizeBoundUniqueVector<Resource> missingResources
     public SizeBoundUniqueVector<Resource> getMissingResources() {
         return mMissingResources;
     }
+
+    public Vector<Resource> getLazyResources() {
+        // todo
```
Comment (ShivamPokhriyal) [inline src/main/java/org/commcare/resources/model/ResourceTable.java:1195]: What's this todo for?

File: src/main/java/org/commcare/xml/ResourceParser.java [hunk 2]
```diff
@@ -46,6 +47,6 @@ public Resource parse() throws InvalidStructureException, IOException, XmlPullPa
             }
         }
 
-        return new Resource(version, id, locations, descriptor);
+        return new Resource(version, id, locations, descriptor, lazy != null && lazy.contentEquals("true"));
```
Comment (ShivamPokhriyal) [inline src/main/java/org/commcare/xml/ResourceParser.java:50]: `"true".contentEquals(lazy)` can be used to avoid null checks.

Comment (shubham1g5) [reply src/main/java/org/commcare/xml/ResourceParser.java:50]: "true".contentEquals(null) does throw a NPE

File: src/main/java/org/commcare/resources/model/ResourceTable.java (hunk 1 above)
Comment (shubham1g5) [reply src/main/java/org/commcare/resources/model/ResourceTable.java:1195]: sorry, left it by mistake

File: src/main/java/org/commcare/xml/ResourceParser.java (hunk 2 above)
Comment (ShivamPokhriyal) [reply src/main/java/org/commcare/xml/ResourceParser.java:50]: Ohh, sorry for that! Just curious why are  we using `contentEquals` over `equals`?

Comment (shubham1g5) [reply src/main/java/org/commcare/xml/ResourceParser.java:50]: no particular reason, they both are same for this use case. `contentEquals` is a more free implementation of `equals` allowing to compare across different `CharSequence` implementations so I just default to it for any string comparison.

File: src/main/java/org/commcare/resources/model/Resource.java [hunk 3]
```diff
@@ -111,7 +112,8 @@
     // Not sure if we want this persisted just yet...
     protected String parent;
 
-    private String descriptor;
+    protected String descriptor;
+    private boolean lazy;
```
Comment (ctsims) [inline src/main/java/org/commcare/resources/model/Resource.java:116]: I think this means we're also going to need to bump the appdb database version in formplayer

File: src/main/java/org/commcare/resources/model/Resource.java [hunk 4]
```diff
@@ -315,13 +332,15 @@ public Object getMetaData(String fieldName) {
                 return parent == null ? "" : parent;
             case META_INDEX_VERSION:
                 return version;
+            case META_INDEX_LAZY:
+                return lazy;
```
Comment (ctsims) [inline src/main/java/org/commcare/resources/model/Resource.java:336]: I'm a little surprised to see this works with a boolean, btw. My intuition is that we should hard code an explicit string value here, and convert it to/from boolean. We've started making soft assumptions elsewhere that all metadata values are strings, and I'm worried that implicit conversions.

File: src/main/java/org/commcare/resources/model/ResourceTable.java [hunk 5]
```diff
@@ -1175,4 +1198,12 @@ public void setMissingResources(SizeBoundUniqueVector<Resource> missingResources
     public SizeBoundUniqueVector<Resource> getMissingResources() {
         return mMissingResources;
     }
+
+    public Vector<Resource> getLazyResources() {
+        return storage.getRecordsForValues(new String[]{Resource.META_INDEX_LAZY}, new Boolean[]{true});
```
Comment (ctsims) [inline src/main/java/org/commcare/resources/model/ResourceTable.java:1203]: do we use this now that we're mostly iterating on ID's? This feels a bit like a potential bomb call, it'll grab a LOT of records.

File: src/main/java/org/javarosa/core/services/storage/IStorageUtilityIndexed.java [hunk 6]
```diff
@@ -207,6 +207,19 @@
...
+     * Retrieves a Vector of Externalizable objects from the storage for which, for each field in fieldNames,
+     * the record has the correct corresponding value in values
+     *
+     * @param metaFieldNames A list of metadata field names to match
+     * @param values     The values which must match the field names provided
+     *
+     * @return A Vector of Externalizable objects e, such that the fields specified are equal to the corresponding values provided.
+     *
+     */
+    Vector<E> getRecordsForValues(String[] metaFieldNames, Object[] values);
```
Comment (ctsims) [inline src/main/java/org/javarosa/core/services/storage/IStorageUtilityIndexed.java:221]: Is this the same signature as the custom implementations in Android AND formplayer? Is good to finally have it committed in code.

**Pull request merged**